src/librustc_mir/transform/const_prop.rs

```diff
@@ -985,8 +1026,48 @@ impl<'mir, 'tcx> MutVisitor<'tcx> for ConstPropagator<'mir, 'tcx> {
             | TerminatorKind::GeneratorDrop
             | TerminatorKind::FalseEdges { .. }
             | TerminatorKind::FalseUnwind { .. } => {}
-            //FIXME(wesleywiser) Call does have Operands that could be const-propagated
-            TerminatorKind::Call { .. } => {}
+            TerminatorKind::Call { ref mut args, .. } => {
+                // Every argument in our function calls can be const propagated.
+                // We try to, before exiting.
+                for opr in args {
+                    // We see if the operand can be evaluated, and if so, we continue.
+                    if let Some(op_ty) = self.eval_operand(&opr, source_info) {
+                        // We must also ask if it *should* be evaluated.
+                        if self.should_const_prop(op_ty) {
+                            if let interpret::Operand::Immediate(immediate) = *op_ty {
+                                // Only Immediate from here on
+                                // FIXME(felix91gr): The code only works for Immediate. Investigate
+                                // if it could be made to work for Indirect as well.
+                                if let interpret::Immediate::Scalar(scalar_mu) = immediate {
+                                    // FIXME(felix91gr): This only works for Scalar
+                                    // Could probably be expanded for Scalar Tuples (inside the Immediate)
+                                    if let ScalarMaybeUndef::Scalar(scalar) = scalar_mu {
+                                        let operand_type = opr.ty(&self.local_decls, self.tcx);
+                                        *opr = self.operand_from_scalar(
+                                            scalar,
+                                            operand_type,
+                                            source_info.span,
+                                        );
+                                    }
+                                }
+                            }
+                        }
+                    }
+                }
+            }
+        }
+        // We remove all Locals which are restricted in propagation to their containing blocks.
+        // We wouldn't need to clone, but the borrow checker can't see that we're not aliasing
+        // the locals_of_current_block field, so we need to clone it first.
+        for local in self.locals_of_current_block.clone().iter() {
```

Comment:
I have an idea for this. We can probably iterate over the locals. Since the Locals aren't anything other than just numbers from `1` to `#locals-1`, we can do something like this to avoid cloning the `BitSet` at every `Terminator` visit.

```rust
for local in 0..self.local_max {
    if self.locals_of_current_block.remove(local) {
        self.remove_const(local);
    }
}
```

This is the function of `BitSet` that we'd be using: https://doc.rust-lang.org/nightly/nightly-rustc/rustc_index/bit_set/struct.BitSet.html#method.remove

@wesleywiser

Reply:
Note: as said in the comments, this `clone` isn't necessary at all; it's just to please the borrow checker because it can't yet see that `remove_const` and `locals_of_current_block` are independent and can therefore be read and written to at the same time without consequences.